CRM/Core/BAO/CMSUser.php

```diff
@@ -144,65 +144,41 @@ public static function formRule($fields, $files, $form) {
 
     $config = CRM_Core_Config::singleton();
 
-    $isDrupal = $config->userSystem->is_drupal;
-    $isJoomla = ucfirst($config->userFramework) == 'Joomla';
-    $isWordPress = $config->userFramework == 'WordPress';
-
     $errors = [];
-    if ($isDrupal || $isJoomla || $isWordPress) {
-      $emailName = NULL;
-      if (!empty($form->_bltID) && array_key_exists("email-{$form->_bltID}", $fields)) {
-        // this is a transaction related page
-        $emailName = 'email-' . $form->_bltID;
-      }
-      else {
-        // find the email field in a profile page
-        foreach ($fields as $name => $dontCare) {
-          if (substr($name, 0, 5) == 'email') {
-            $emailName = $name;
-            break;
-          }
-        }
-      }
 
-      if ($emailName == NULL) {
-        $errors['_qf_default'] = ts('Could not find an email address.');
-        return $errors;
-      }
+    $emailName = $config->userSystem->getEmailFieldName($form, $fields);
 
-      if (empty($fields['cms_name'])) {
-        $errors['cms_name'] = ts('Please specify a username.');
-      }
+    $params = [
+      'name' => $fields['cms_name'],
+      'mail' => isset($fields[$emailName]) ? $fields[$emailName] : '',
+    ];
 
-      if (empty($fields[$emailName])) {
-        $errors[$emailName] = ts('Please specify a valid email address.');
-      }
+    // Set generic errors messages.
+    if ($emailName == '') {
+      $errors['_qf_default'] = ts('Could not find an email address.');
+    }
 
-      if ($config->userSystem->isPasswordUserGenerated()) {
-        if (empty($fields['cms_pass']) ||
-          empty($fields['cms_confirm_pass'])
-        ) {
-          $errors['cms_pass'] = ts('Please enter a password.');
-        }
-        if ($fields['cms_pass'] != $fields['cms_confirm_pass']) {
-          $errors['cms_pass'] = ts('Password and Confirm Password values are not the same.');
-        }
-      }
+    if (empty($params['name'])) {
+      $errors['cms_name'] = ts('Please specify a username.');
+    }
 
-      if (!empty($errors)) {
-        return $errors;
-      }
```

Comment:
See my comment on the other PR. Removing/moving this to the bottom is problematic because checkUserNameEmailExists effectively clears any previously set errors. But I think the real fix for this would be in checkUserNameEmailExists. In general I like the cleanup to remove this mysterious early return.

Reply:
I think I've addressed it by having Drupal 7 and Backdrop not reset the errors each time.